### PR TITLE
Add image-flash, gateware-flash, etc support for pano_logic_g2 target.

### PR DESCRIPTION
**Pull request by skiphansen**
I also adjusted the flash partitioning as I discovered that I had the wrong erase size before.  The minimum erase size is 256MB, not 128MB.

Hopefully the changes to the top level Make file will be useful for other targets.

### CODE REVIEW

**Comment by skiphansen** [comment]
The build for C=vexriscv TC="vivado" P=arty T="base" F=micropython failed.  I'm looking into it.

**Comment by mithro** [comment]
```
 OBJCOPY  firmware.bin
chmod -x firmware.bin
python -m litex.soc.software.mkmscimg -f --little firmware.bin -o firmware.fbi
make[1]: Leaving directory '/home/travis/build/timvideos/litex-buildenv/build/arty_base_vexriscv/software/firmware'
real	0m4.117s
user	0m3.585s
sys	0m0.560s
python mkimage.py \
	   \
	--override-gateware=build/arty_base_vexriscv//gateware/top.bin \
	--override-bios=build/arty_base_vexriscv//software/bios/bios.bin \
	--override-firmware=build/arty_base_vexriscv//software/micropython/firmware.fbi \
	--output-file=build/arty_base_vexriscv//image-gateware+bios+micropython.bin
Traceback (most recent call last):
  File "mkimage.py", line 261, in <module>
    main()
  File "mkimage.py", line 160, in main
    "Use --override-firmware=none for no firmware." % firmware)
AssertionError: Firmware file 'build/arty_base_vexriscv//software/micropython/firmware.fbi' not found! Use --override-firmware=none for no firmware.
Makefile:233: recipe for target 'build/arty_base_vexriscv//image-gateware+bios+micropython.bin' failed
make: *** [build/arty_base_vexriscv//image-gateware+bios+micropython.bin] Error 1
```

**Comment by mithro** [comment]
@mateusz-holenko -- I'll leave you to review and merge.

**Comment by skiphansen** [comment]
Oh well.  The build that failed yesterday is fixed but there's a new failure.  I'll look into it ASAP.

File: Makefile [hunk 1]
```diff
@@ -212,6 +212,23 @@ FIRMWARE_FBI=$(FIRMWARE_FILEBASE).fbi
...
 endif
 
+ifeq ($(FIRMWARE),linux)
+# Linux Image component files - kernel+emulator+dts+rootfs
+    define fbi_rule
+        $(1): $(subst .fbi,,$(1))
+    endef
+    
+    OVERRIDE_FIRMWARE=--override-firmware=none
+    BUILDROOT_IMAGES = third_party/buildroot/output/images
```
[the remark this answers is not in the document]
Comment (skiphansen) [reply Makefile:222]: I didn't know BUILD_BUILDROOT was optional.  I was following the [build instructions](https://github.com/timvideos/litex-buildenv/wiki/HowTo-Linux-on-Pano-Logic-G2) which said to set it.  I'll have a look.

File: targets/pano_logic_g2/Makefile.mk [hunk 2]
```diff
@@ -12,26 +12,56 @@ COMM_PORT ?= /dev/ttyUSB0
...
 
-gateware-flash-$(PLATFORM):
-	@echo "Unsupported"
-	@false
+ifeq ($(FIRMWARE),linux)
+GATEWARE_BIN = $(TARGET_BUILD_DIR)/gateware+emulator+dtb.bin
+
+$(GATEWARE_BIN): $(GATEWARE_FILEBASE).bin $(DTB_FBI) $(EMULATOR_FBI)
+	# note: emulator and DTB are flash with gateware to save flash space
+	dd if=$(GATEWARE_FILEBASE).bin of=$@ bs=4259840 conv=sync
```
[the remark this answers is not in the document]
Comment (skiphansen) [reply targets/pano_logic_g2/Makefile.mk:26]: Yes that is correct and I agree.

**Comment by skiphansen** [comment]
Please rerun the C=picorv32.minimal TC="vivado" P=arty T="base net" build.  It passed here, it's not clear to me what failed ... temp network issue perhaps?

**Comment by mithro** [comment]
@skiphansen Reran the failing tests.

File: targets/pano_logic_g2/Makefile.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply targets/pano_logic_g2/Makefile.mk:30]: It looks like there is no support for generating `Makefile.inc` with constants in buildenv at the moment. 

I plan to look at it soon, but this shouldn't block this PR.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (skiphansen) [reply Makefile]: Suggested change made, tested with BUILD_BUILDROOT only.  I'm not sure how to run a non-buildroot image, nfs root perhaps?

File: targets/pano_logic_g2/Makefile.mk [hunk 3]
```diff
@@ -11,27 +11,84 @@ TARGET ?= $(DEFAULT_TARGET)
...
 
+ifeq ($(FIRMWARE),linux)
+
+
+OVERRIDE_FIRMWARE=--override-firmware=none
+define fbi_rule
+$(1): $(subst .fbi,,$(1))
+endef
+
+TARGET_BUILD_DIR = build/$(FULL_PLATFORM)_$(TARGET)_$(FULL_CPU)
```
Comment (mateusz-holenko) [inline targets/pano_logic_g2/Makefile.mk:22]: this has already been defined in `Makefile:144`; is it needed again?

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply Makefile]: Running `./scrips/build-linux.sh` with or without `BUILD_BUILDROOT` should result in generating the same set of four binaries:
```
<build_dir>/software/linux/firmware.bin
<build_dir>/software/linux/riscv32-rootfs.cpio
<build_dir>/software/linux/rv32.dtb
<build_dir>/emulator/emulator.bin
```
(They are just copied/symlinked from the buildroot output directory if `BUILD_BUILDROOT` is set).
When using images from this location you don't even need to know if they were built by BUILDROOT or not.

I see however that you are referencing `.fbi` files. What are those? Are they generated by buildroot? If so than maybe you are right that you need BUILDROOT-based version.

File: targets/pano_logic_g2/Makefile.mk (hunk 3 above)
Comment (skiphansen) [reply targets/pano_logic_g2/Makefile.mk:22]: Yes it is.  The Makefile.mk is included on line 120 of the top level make file so TARGET_BUILD_DIR isn't defined when it's included.  Additionally TARGET is defined my the Makefile.mk so TARGET_BUILD_DIR  can't be defined by the top level Makefile before the target makefile is included.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (skiphansen) [reply Makefile]: Sorry, I missed that one I'll change it.

FBI files (flash binary image perhaps?) are what the BIOS expected to be in flash.  The are just the binary with a header that has the length and CRC.  If the binary is programmed instead the BIOS will error out.

As far as I know .fbi files are specific to LiteX, they have nothing to do with buildroot.

File: targets/pano_logic_g2/Makefile.mk (hunk 3 above)
Comment (mateusz-holenko) [reply targets/pano_logic_g2/Makefile.mk:22]: That's fine then

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply Makefile]: Looks good now.

**Pull request merged**